File: swiftReader/swiftReader/ParseWebSite.m
//
//  ParseWebSite.m
//  EReader
//
//  Created by 姚小勇 on 15/11/29.
//  Copyright © 2015年 姚小勇. All rights reserved.
//

#import "ParseWebSite.h"
#import "TFHpple.h"


@implementation ParseWebSite
static ParseWebSite *instance;
+ (instancetype) sharedInstance
{
    static dispatch_once_t onceToken;
    dispatch_once(&onceToken, ^{
        instance = [[self alloc] init];
    });
    return instance;
}

//////////////////////////////////////////////////////////////////////////////////
-(TFHpple*)THppleParseURL:(NSString*)keyValue
{
    NSString *requestURLwithParams = [NSString stringWithFormat:@"%@%@",QINGDISEARCHARTICLENAMEBASE, keyValue];

    NSURL *url = [NSURL URLWithString:[requestURLwithParams stringByAddingPercentEscapesUsingEncoding:CHINESE_CODING]];
    //NSLog(@"url1=%@",url);
    NSError *err = nil;

    NSString *htmlString = [NSString stringWithContentsOfURL:url
                                                    encoding:CHINESE_CODING
                                                       error:&err];
    //NSLog(@"htmlString:%@", htmlString);
    TFHpple *xpathParser = [[TFHpple alloc] initWithHTMLData:[htmlString dataUsingEncoding:NSUnicodeStringEncoding]];
    return xpathParser;
}

-(void)parseANovelIndex:(TFHpple *)xpathParser key:(NSString*)keyValue{
    NSArray *items = [xpathParser searchWithXPathQuery:@"//div[@class='articleInfo']"];
    if(items.count == 0){
        return;
    }
    TFHppleElement *hppleElement = items[0];
    NSString *novelIndexURL     = [[[[hppleElement firstChildWithClassName:@"articleInfoLeft"] firstChildWithTagName:@"p"] firstChildWithTagName:@"a"] objectForKey:@"href"];
    NSString *novelIndexJPGURL  = [[[[[[hppleElement firstChildWithClassName:@"articleInfoLeft"] firstChildWithTagName:@"p"] firstChildWithTagName:@"a"] firstChildWithTagName:@"img"] attributes] objectForKey:@"src"];
    NSString *novelName         = [[[[[[hppleElement firstChildWithClassName:@"articleInfoLeft"] firstChildWithTagName:@"p"] firstChildWithTagName:@"a"] firstChildWithTagName:@"img"] attributes] objectForKey:@"alt"];
    NSString *novelRitle        = [[[[[[hppleElement firstChildWithClassName:@"articleInfoLeft"] firstChildWithTagName:@"p"] firstChildWithTagName:@"a"] firstChildWithTagName:@"img"] attributes] objectForKey:@"title"];
    NSString *novelSummary      = [[[[[hppleElement firstChildWithClassName:@"articleInfoRight"] firstChildWithTagName:@"span"] firstChildWithTagName:@"dl"] firstChildWithTagName:@"dd"] text];
    NSString *novelAuthor       = [[[[[hppleElement firstChildWithClassName:@"articleInfoRight"] firstChildWithTagName:@"span"] firstChildWithTagName:@"h1"] firstChildWithTagName:@"b"] text];
    NSMutableArray *array = [NSMutableArray array];
    //NSLog(@"书名000=%@",     novelName);
    if([novelName containsString:@"："]){
        array = [novelName componentsSeparatedByString:@"："];
    }else{
        array = [novelName componentsSeparatedByString:@":"];
    }
    novelName = [array lastObject];

    if([novelAuthor containsString:@"："]){
        array = [novelAuthor componentsSeparatedByString:@"："];
    }else{
        array = [novelAuthor componentsSeparatedByString:@":"];
    }
    novelAuthor = [array lastObject];

//    NSLog(@"链接==%@",    novelIndexURL);
//    NSLog(@"封页=%@",     novelIndexJPGURL);
//    NSLog(@"书名=%@",     novelName);
//    NSLog(@"标题=%@",     novelRitle);
//    NSLog(@"简介=%@",     novelSummary);
//    NSLog(@"作者=%@",     novelAuthor);
    NSLog(@"================================================================");

    searchResult *result = [[searchResult alloc]init];
    result.searchkey = keyValue;
    result.novelName = novelName;
    result.novelIndexURL = novelIndexURL;
    result.novelJPGURL = novelIndexJPGURL;
    result.novelAuthor = novelAuthor;
    result.novelSummary = novelSummary;
    dispatch_async(dispatch_get_global_queue(0, 0), ^{
        NSArray *array = [searchResult findByCriteria:[NSString stringWithFormat:@" WHERE searchkey ='%@' and novelName = '%@'", keyValue, novelName]];
        if(array.count > 0)
        {
            DDLog();
            dispatch_async(dispatch_get_main_queue(), ^{
                DDLog();
                [[searchViewController sharedInstance] reloadDataSource];
            });
        }else{
            DDLog(@"save searchResult");
            [result save];
            dispatch_async(dispatch_get_main_queue(), ^{
                DDLog();
                [[searchViewController sharedInstance] reloadDataSource];
            });

        }
    });
}

//////////////////////////////////////////////////////////////////////////////////
-(void)searchWithKey:(NSString*)keyValue
      withReturnValueHandleBlock: (ReturnValueHandleBlock)returnHandleBlock
            withErrorHandleBlock: (ErrorHandleBlock)errorHandleBlock
          withFailureHandleBlock: (FailureHandleBlock)failureHandleBlock
{
    [[NSUserDefaults standardUserDefaults] setObject:keyValue forKey:@"searchkey"];
    NSMutableArray *searchResultArray = [NSMutableArray array];
    NSArray *array = [searchResult findByCriteria:[NSString stringWithFormat:@" WHERE searchkey ='%@'", keyValue]];
    if(array.count > 0){
        DDLog(@"有搜索记录");
        dispatch_async(dispatch_get_main_queue(), ^{
            DDLog();
            [[searchViewController sharedInstance] reloadDataSource];
        });
    }else{
        TFHpple *xpathParser = [self THppleParseURL:keyValue];

        NSArray *items = [NSMutableArray array];
        DDLog();
        items = [xpathParser searchWithXPathQuery:@"//ul[@class='info']"];
        if([items count] > 0){
            DDLog();
            for(TFHppleElement *hppleElement in items){
                NSError *err = nil;
                DDLog();
                NSString *htmlString = [NSString stringWithContentsOfURL:[NSURL URLWithString:[[hppleElement firstChildWithTagName:@"a"] objectForKey:@"href"]]
                                                                encoding:CHINESE_CODING
                                                                   error:&err];
                //NSLog(@"htmlString:%@", htmlString);
                TFHpple *xpathParser2 = [[TFHpple alloc] initWithHTMLData:[htmlString dataUsingEncoding:NSUnicodeStringEncoding]];
                [self parseANovelIndex:xpathParser2 key:keyValue];

            }
        }else{
            DDLog();
            [self parseANovelIndex:xpathParser key:keyValue];
        }
    }
}

////////////////////////////////////////////////////////////////////////////////////
-(void) downloadNovel:(NSDictionary*)dic{
    NSError *err = nil;
    NSString *novelName     =   [dic valueForKey:OBJ_NOVELNAME];
    NSString *novelIndexURL =   [dic valueForKey:OBJ_NOVELINDEXURL];

    novel *book             =   [[novel alloc]init];
    book.novelName          =   novelName;
    book.novelIndexURL      =   [dic valueForKey:OBJ_NOVELINDEXURL];
    book.novelSummary       =   [dic valueForKey:NOVELSUMMARY];
    book.novelJPGURL        =   [dic valueForKey:OBJ_NOVELJPGURL];
    book.novelAuthor        =   [dic valueForKey:OBJ_NOVELAUTHOR];

    dispatch_async(dispatch_get_global_queue(0, 0), ^{
        [book save];
        NSArray *array = [novel findByCriteria:[NSString stringWithFormat:@" WHERE novelName = '%@'", novelName]];
        if(array.count > 0){
        }else{
            DDLog(@"save novel");
            [book save];
        }
    });

    NSString *htmlString = [NSString stringWithContentsOfURL:[NSURL URLWithString:novelIndexURL]
                                                    encoding:CHINESE_CODING
                                                       error:&err];
    //NSLog(@"htmlString:%@", htmlString);

    TFHpple *xpathParser = [[TFHpple alloc] initWithHTMLData:[htmlString dataUsingEncoding:NSUnicodeStringEncoding]];

    NSArray *data                    = [xpathParser searchWithXPathQuery:@"//tr//td//a"];
    NSLog(@"[data.count]=%lu", (unsigned long)data.count);
    NSInteger __block downloadedNum = 0;

    NSInteger __block totalChapterNum = data.count;
    for (NSInteger i = 0; i < totalChapterNum; i++){
        TFHppleElement *element = [data objectAtIndex:i];
        dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0), ^{
            NSString *chapterNovelName = [[element objectForKey:@"href"] stringByReplacingOccurrencesOfString:@"html" withString:@"txt"];
            NSString *chapterURL = [novelIndexURL stringByReplacingOccurrencesOfString:@"index.html" withString:[element objectForKey:@"href"]];
            NSURLRequest *request=[NSURLRequest requestWithURL:[NSURL URLWithString:chapterURL]];
            NSURLSessionDataTask *task = [[NSURLSession sharedSession] dataTaskWithRequest:request completionHandler:^(NSData *data, NSURLResponse *response, NSError *error) {
                NSHTTPURLResponse *httpResponse = (NSHTTPURLResponse*)response;
                //NSLog(@" http response : %ld",(long)httpResponse.statusCode);
                if(200 != httpResponse.statusCode){
                }else{
                    NSString *aString = [[NSString alloc] initWithData:data encoding:CHINESE_CODING];
                    //NSLog(@"aString = %@", aString);
                    TFHpple *xpathParser                = [[TFHpple alloc] initWithHTMLData:[aString dataUsingEncoding:NSUnicodeStringEncoding]];
                    NSArray *elements                   = [xpathParser searchWithXPathQuery:@"//div[@class='readerTitle']"]; // get the title
                    if(elements.count == 0){
                        downloadedNum++;
                        NSLog(@"chapterURL=%@", chapterURL);
                        return;
                    }else{
                        TFHppleElement *hppleElement    = elements[0];
                        NSString *chapterTitle          = [[hppleElement firstChildWithTagName:@"h1"] text];
                        NSString *chapterContent        = [self getNovelContentFromChaptrURL:chapterURL];

                        chapter *newchapter             = [[chapter alloc]init];
                        newchapter.novelName            = novelName;
                        newchapter.chapterName          = chapterNovelName;
                        newchapter.chapterURL           = chapterURL;
                        newchapter.chapterContent       = chapterContent;//@"内容是不是太长了，怎么没有存下来";
                        newchapter.chapterTitle         = chapterTitle;
                        dispatch_async(dispatch_get_global_queue(0, 0), ^{
                            NSArray *array = [chapter findByCriteria:[NSString stringWithFormat:@" WHERE chapterName ='%@' and novelName = '%@'", chapterNovelName, novelName]];
                            if(array.count > 0)
                            {
//                                DDLog(@"update chapter");
//                                [newchapter update];
                            }else{
                                [newchapter save];
                            }
                        });
                    }
                }

                downloadedNum++;
                if(totalChapterNum == (downloadedNum +1)){
                    //DDLog(@"allkeyss=%@", [dic allKeys]);
                    NSMutableDictionary *novelDic = [NSMutableDictionary dictionaryWithDictionary:dic];
                    NSMutableDictionary *downloadOK = [NSMutableDictionary dictionary];
                    [downloadOK setObject:@"downloadOK" forKey:@"action"];
                    [[NotificationCenter sharedInstance] sendNotification:downloadOK];
                }

                printf("%ld  ",(long)downloadedNum);
            }];

            [task resume];
        });
    }
}


-(NSString *)getNovelContentFromChaptrURL:(NSString*)chapterURL
{
    NSMutableArray *array = [[NSMutableArray alloc] initWithObjects:@"", nil];
    NSString *htmlString = [NSString stringWithContentsOfURL:[NSURL URLWithString:chapterURL]
                                                    encoding:CHINESE_CODING
                                                       error:nil];
    NSData *htmlData = [htmlString dataUsingEncoding:CHINESE_CODING];
    TFHpple *xpathParser = [[TFHpple alloc] initWithHTMLData:htmlData];
    NSArray *elements  = [xpathParser searchWithXPathQuery:@"//div"]; // get the title

    for(TFHppleElement *hppleElement in elements){
        if ([[hppleElement objectForKey:@"id"] isEqualToString:@"content"]) {
            for (TFHppleElement* child in hppleElement.children){
                if ([child isTextNode]){
                    [array addObject:child.content];
                }
            }
        }
    }
    NSString *chapterContent = [array componentsJoinedByString:@" "];
    return chapterContent;
}

@end

File: swiftReader/swiftReader/ParseWebsite.swift
//
//  ParseWebsite.swift
//  swiftReader
//
//  Created by 姚小勇 on 15/12/1.
//  Copyright © 2015年 姚小勇. All rights reserved.
//

import UIKit

import Alamofire
import Async

class ParseWebsite: NSObject {

    class var sharedInstance : ParseWebsite {
        struct Static {
            static let instance : ParseWebsite = ParseWebsite()
        }
        return Static.instance
    }

    func parseANovelIndex(xpathParser:TFHpple, keyValue:NSString){
        var items = xpathParser.searchWithXPathQuery("//div[@class='articleInfo']")
        print("========>item1=\(items.count)")
        if(items.count == 0){
            //var items1 = xpathParser.searchWithXPathQuery("//div")
            return
        }else{
            var hppleElement  = items[0]
            let novelIndexURL = hppleElement.firstChildWithClassName("articleInfoLeft").firstChildWithTagName("p").firstChildWithTagName("a").objectForKey("href")
            let nodevalue     = hppleElement.firstChildWithClassName("articleInfoLeft").firstChildWithTagName("p").firstChildWithTagName("a").firstChildWithTagName("img").attributes as NSDictionary
            let jpgurl:String = nodevalue.valueForKey("src") as! String

            let novelJPGURL   = NSData(contentsOfURL:NSURL(string: jpgurl)!)

            var novelName     = nodevalue.valueForKey("alt")
            let novelTitle    = nodevalue.valueForKey("title")
            let novelSummary  = hppleElement.firstChildWithClassName("articleInfoRight").firstChildWithTagName("span").firstChildWithTagName("dl").firstChildWithTagName("dd").text()
            let novelAuthor   = hppleElement.firstChildWithClassName("articleInfoRight").firstChildWithTagName("span").firstChildWithTagName("h1").firstChildWithTagName("b").text()

            let result123 = novelName?.rangeOfString(":", options: .LiteralSearch)
            print("====rsult123======>\(result123)")

            if((novelName?.containsString("：")) != nil){
                let array = (novelName?.componentsSeparatedByString("："))! as NSArray
                novelName = array.lastObject as! String
            }
            if((novelName?.containsString(":")) != nil){
                let array = (novelName?.componentsSeparatedByString(":"))! as NSArray
                novelName = array.lastObject as! String
            }
            let result = searchResult()
            result.searchkey        = keyValue as String
            result.novelName        = novelName as! String
            result.novelIndexURL    = novelIndexURL
            result.novelJPGURL      = novelJPGURL
            result.novelSummary     = novelSummary
            result.novelAuthor      = novelAuthor

            NSLog("链接==\(novelIndexURL)")
            NSLog("title==\(novelTitle)");
            //NSLog("封页=\(novelJPGURL)")
            NSLog("书名=\(novelName)")
            NSLog("简介=\(novelSummary)")
            NSLog("作者=\(novelAuthor)")
            dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0), {
                result.save()
                NSLog("%@", result)
                dispatch_async(dispatch_get_main_queue(), {
                    print("send a notofication")
                    NSNotificationCenter.defaultCenter().postNotificationName("MyMotification", object: self, userInfo: ["action":"searchOK"])
                })
            })

        }

    }
    /////////////////////////////////////////////////////////////////////////////////////////////////////////////////
    func searchWithKey(keyValue:String){
        let array = searchResult.findByCriteria(" WHERE searchkey ='\(keyValue)'") as NSArray
        if(array.count > 0){
            //NSLog("有搜索记录:\(array)")
            dispatch_async(dispatch_get_main_queue(), {
                print("send a notofication")
                NSNotificationCenter.defaultCenter().postNotificationName("MyMotification", object: self, userInfo: ["action":"searchOK"])
                searchViewController.sharedInstance.reloadDataSource()
            })
        }else{
            NSLog("没有搜索记录")
            Alamofire.request(.GET, utils.ParseWithKey(keyValue), parameters: nil).response {
                (request, response, data, error) in
                //print(response)
                let hppleParse = TFHpple(HTMLData: data)
                let array = hppleParse.searchWithXPathQuery("//ul[@class='info']")
                if(array.count > 0){
                    print("有结果，说明吧")
                    for item in array{
                        let urlstr = item.firstChildWithTagName("a").objectForKey("href")
                        let parse = utils.THppleParseWithURL(urlstr)
                        self.parseANovelIndex(parse, keyValue: keyValue)
//                        Alamofire.request(.GET, urlstr, parameters: nil).response {
//                            (request, response, string, error) in
//                            print(urlstr)
//                            print(response)
//                            let parse = TFHpple(HTMLData: data)
//                            self.parseANovelIndex(parse, keyValue: keyValue)
//                        }
                    }
                }
                else{
                    print("没结果，说明吧")
                    self.parseANovelIndex(hppleParse, keyValue: keyValue)
                }

            }

        }
    }

    func downloadNovel(book:novel){
        dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0), {
            let array = novel.findByCriteria(" WHERE novelName = '\(book.novelName)'")
            if(array.count == 0){
                book.save()
            }
        })

        //let hppleParse : TFHpple =  utils.THppleParseWithURL(book.novelIndexURL)
        Alamofire.request(.GET, book.novelIndexURL, parameters: nil).response {
            (request, response, data, error) in
            let hppleParse = TFHpple(HTMLData: data)
            let array = hppleParse.searchWithXPathQuery("//tr//td//a")

            var downloadNum = 0
            let totalChapterNum = array.count

            Async.background {
                for var index = 0; index < totalChapterNum; index++ {
                    let hppleElement:TFHppleElement = array[index] as! TFHppleElement
                    let chapterName = hppleElement.objectForKey("href")
                    let chapterURL = book.novelIndexURL.stringByReplacingOccurrencesOfString("index.html", withString: chapterName)
                    Alamofire.request(.GET, chapterURL, parameters: nil).response {
                        (request, response, data, error) in

                        downloadNum++

                        let chapterParse: TFHpple = TFHpple.init(HTMLData: data)

                        let elements = chapterParse.searchWithXPathQuery("//div[@class='readerTitle']")
                        if(elements.count == 0){
                            //print("=======1>\(request)-------\(response)")
                            return
                        }else if(elements == nil){
                            //print("=======3>\(data)")
                        }
                        let chapterTitle = elements[0].firstChildWithTagName("h1").text()

                        let contentElement = chapterParse.searchWithXPathQuery("//div[@class='readerContent']//div[@id='content']")
                        if(contentElement.count == 0){
                            print("=======2>\(chapterURL)")
                            return
                        }else if(contentElement == nil){
                            print("=======4>\(chapterURL)")
                        }
                        let array = NSMutableArray()
                        for hppleElement in contentElement{
                            for child in hppleElement.children as NSArray{
                                if (child.isTextNode()){
                                    array.addObject(child.content)
                                }
                            }
                        }
                        let chapterContent = array.componentsJoinedByString(" ")


                        print("\(chapterTitle)-----\(downloadNum)-----\(totalChapterNum)")

                        let oneChapter = chapter()
                        oneChapter.novelName = book.novelName
                        oneChapter.chapterURL = chapterURL
                        oneChapter.chapterName = chapterName
                        oneChapter.chapterTitle = chapterTitle
                        oneChapter.chapterContent = chapterContent
                        let findResult = chapter.findByCriteria(" WHERE chapterName ='\(chapterName)' and novelName = '\(book.novelName)'")
                        if(findResult.count == 0){
                            oneChapter.save()
                        }
                    }
                }
                }.main {
                    print("send a notofication")
                    NSNotificationCenter.defaultCenter().postNotificationName("MyMotification", object: self, userInfo: ["action":"downloadOK"])
            }

        }
    }



}
